To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

### News

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1" huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [2]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 15 trillion tokens model 2x faster!
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # We also uploaded 4bit for 405b!
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # New Mistral 12b 2x faster!
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/Meta-Llama-3.1-8B",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.5.2: Fast Llama patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors:   0%|          | 0.00/5.96G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/235 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/50.6k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/459 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [3]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.5.2 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


In [4]:
import torch
import json
import sentencepiece as spm
from datasets import Dataset
from transformers import AutoTokenizer
from google.colab import files
import os


In [5]:

# 1. Upload your files to Colab
def upload_files_to_colab():
    print("Please upload your SentencePiece model file (.model)...")
    sp_model_file = files.upload()
    sp_model_path = list(sp_model_file.keys())[0]

    print("Please upload your Sindhi tokens JSON file...")
    tokens_file = files.upload()
    tokens_path = list(tokens_file.keys())[0]

    print("Please upload your training data file...")
    data_file = files.upload()
    data_path = list(data_file.keys())[0]

    return sp_model_path, tokens_path, data_path

In [6]:

# 2. Load your Sindhi SentencePiece tokenizer
def load_sindhi_tokenizer(sp_model_path):
    sp_processor = spm.SentencePieceProcessor()
    sp_processor.load(sp_model_path)
    print(f"Loaded SentencePiece model with {sp_processor.get_piece_size()} pieces")
    return sp_processor


In [7]:
# 3. Load your Sindhi tokens
def load_sindhi_tokens(tokens_json_path):
    with open(tokens_json_path, 'r', encoding='utf-8') as f:
        sindhi_tokens = json.load(f)
    print(f"Loaded {len(sindhi_tokens)} Sindhi tokens")
    return sindhi_tokens


In [8]:
# 4. Prepare your dataset
def prepare_sindhi_dataset(data_file_path):
    """
    Prepare your Sindhi dataset for fine-tuning.
    """
    # Determine file type
    if data_file_path.endswith('.json'):
        with open(data_file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
    elif data_file_path.endswith('.txt'):
        with open(data_file_path, 'r', encoding='utf-8') as f:
            data = [line.strip() for line in f.readlines() if line.strip()]
    else:
        raise ValueError("Unsupported file format. Please provide JSON or TXT file.")



    # Format data
    formatted_data = []

    # If data is a list of strings (text samples)
    if isinstance(data, list) and all(isinstance(item, str) for item in data):
        for text in data:
            formatted_data.append({"text": text})

    # If data is a list of dictionaries
    elif isinstance(data, list) and all(isinstance(item, dict) for item in data):
        for item in data:
            if "prompt" in item and "completion" in item:
                # Format as completion task
                formatted_data.append({
                    "prompt": item["prompt"],
                    "completion": item["completion"]
                })
            elif "text" in item:
                # Format as text-only task
                formatted_data.append({"text": item["text"]})

    # Create HuggingFace dataset
    dataset = Dataset.from_list(formatted_data)
    print(f"Dataset prepared with {len(dataset)} examples")
    print(f"Dataset format: {dataset[0].keys()}")
    return dataset

In [12]:


# 5. Load base model tokenizer and merge Sindhi tokens
def prepare_tokenizer(base_model_name, sindhi_tokens=None):
    # Load the base model tokenizer
    tokenizer = AutoTokenizer.from_pretrained(base_model_name)
    print(f"Loaded base tokenizer: {tokenizer.name_or_path}")

    # If you need to incorporate your Sindhi tokens into the base tokenizer
    if sindhi_tokens is not None:
        print("Adding Sindhi tokens to the base tokenizer...")
        # Get current vocabulary size
        orig_vocab_size = len(tokenizer)

        # Add Sindhi tokens to the base tokenizer
        new_tokens = [token for token in sindhi_tokens if token not in tokenizer.get_vocab()]
        if new_tokens:
            tokenizer.add_tokens(new_tokens)
            print(f"Added {len(new_tokens)} new Sindhi tokens to the tokenizer")
            print(f"Vocabulary size increased from {orig_vocab_size} to {len(tokenizer)}")

    return tokenizer

# Main execution code for loading and dataset preparation only
if __name__ == "__main__":
    # 1. Upload necessary files
    print("Starting file uploads...")
    sp_model_path, tokens_path, data_path = upload_files_to_colab()

    # 2. Load base model name
    BASE_MODEL_NAME = "unsloth/Meta-Llama-3.1-8B"  # Change as needed

    # 3. Load Sindhi tokenizer and tokens
    sindhi_sp_tokenizer = load_sindhi_tokenizer(sp_model_path)
    sindhi_tokens = load_sindhi_tokens(tokens_path)

    # 4. Prepare dataset
    dataset = prepare_sindhi_dataset(data_path)

# 5. Load base model tokenizer and merge Sindhi tokens
def prepare_tokenizer(base_model_name, sindhi_tokens=None):
    # Load the base model tokenizer
    tokenizer = AutoTokenizer.from_pretrained(base_model_name)
    print(f"Loaded base tokenizer: {tokenizer.name_or_path}")

    # If you need to incorporate your Sindhi tokens into the base tokenizer
    if sindhi_tokens is not None:
        print("Adding Sindhi tokens to the base tokenizer...")
        # Get current vocabulary size
        orig_vocab_size = len(tokenizer)

        # Add Sindhi tokens to the base tokenizer
        # The change is on this line: extract the actual token strings from the dictionaries
        new_tokens = [token['encoded_tokens'] for token in sindhi_tokens if token['encoded_tokens'] not in tokenizer.get_vocab()]
        if new_tokens:
            tokenizer.add_tokens(new_tokens)
            print(f"Added {len(new_tokens)} new Sindhi tokens to the tokenizer")
            print(f"Vocabulary size increased from {orig_vocab_size} to {len(tokenizer)}")

    return tokenizer

Starting file uploads...
Please upload your SentencePiece model file (.model)...


Saving 87M_sindhi_tokenizer 1.model to 87M_sindhi_tokenizer 1 (3).model
Please upload your Sindhi tokens JSON file...


Saving 5_final_final.checked_tokenized.json to 5_final_final.checked_tokenized (1).json
Please upload your training data file...


Saving sampledecoded_text.txt to sampledecoded_text (1).txt
Loaded SentencePiece model with 5000 pieces
Loaded 167 Sindhi tokens
Dataset prepared with 167 examples
Dataset format: dict_keys(['text'])


<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [13]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported

trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    dataset_text_field = "text",
    max_seq_length = max_seq_length,
    dataset_num_proc = 2,
    packing = False, # Can make training 5x faster for short sequences.
    args = TrainingArguments(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        # num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 60,
        learning_rate = 2e-4,
        fp16 = not is_bfloat16_supported(),
        bf16 = is_bfloat16_supported(),
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use this for WandB etc
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/167 [00:00<?, ? examples/s]

In [14]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
7.135 GB of memory reserved.


In [15]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 167 | Num Epochs = 3 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 41,943,040/8,000,000,000 (0.52% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [16]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

653.8793 seconds used for training.
10.9 minutes used for training.
Peak reserved memory = 7.764 GB.
Peak reserved memory for training = 0.629 GB.
Peak reserved memory % of max memory = 52.669 %.
Peak reserved memory for training % of max memory = 4.267 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

**[NEW] Try 2x faster inference in a free Colab for Llama-3.1 8b Instruct [here](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [19]:


# Enable faster inference if available
FastLanguageModel.for_inference(model)  # Enable native 2x faster inference

sindhi_prompt = """ھيٺ ڏنل متن جاري ڪريو.
ان پٽ: {input}
پيداوار: {output}"""

inputs = tokenizer(
    [
        sindhi_prompt.format(
            input="هڪڙو ڇوڪرو جنگل ۾ وڃي رهيو هو.",  # Input sentence to continue
            output="",  # Leave blank for generation
        )
    ],
    return_tensors="pt"
).to("cuda")

outputs = model.generate(**inputs, max_new_tokens=64)
generated_text = tokenizer.batch_decode(outputs)[0]
print(generated_text)

<|begin_of_text|>ھيٺ ڏنل متن جاري ڪريو.  
ان پٽ: هڪڙو ڇوڪرو جنگل ۾ وڃي رهيو هو.  
پيداوار: ڇوڪرن جي پيداوار ڪرڻ ۾ ڪيتري ڳالهيون آهن، اُن ڳالهين کي سڏڻ لاءِ هيٺين لفظن کي استعمال 


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [20]:
model.save_pretrained("llama8bsindhi")  # Local saving
tokenizer.save_pretrained("llama8bsindhitoken")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('llama8bsindhitoken/tokenizer_config.json',
 'llama8bsindhitoken/special_tokens_map.json',
 'llama8bsindhitoken/tokenizer.json')

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [23]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "llama8bsindh", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference


inputs = tokenizer(
    [
        sindhi_prompt.format(
            input="هڪڙو ڇوڪرو جنگل ۾ وڃي رهيو هو.",  # Input sentence to continue
            output="",  # Leave blank for generation
        )
], return_tensors = "pt").to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(**inputs, streamer = text_streamer, max_new_tokens = 128)

<|begin_of_text|>ھيٺ ڏنل متن جاري ڪريو.  
ان پٽ: هڪڙو ڇوڪرو جنگل ۾ وڃي رهيو هو.  
پيداوار: ڪڇ جي ڪن پٽن ۾ ٿر ۽ ٿاريءَ جي فصلون ڪلرن ۾ ڪٽيون وڃن ٿيون، جنهن جي ڪري ٿر ۽ ٿاريءَ جي فصلون جي پيداوار ڊسمبر ۾ شروع ٿي وڃي ٿي، جنهن جي پيداوار کي ڪلرن ۾ ڪٽي ڪري


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
